modules/meshrenderinggl/src/processors/meshrasterizer.cpp

```diff
@@ -1,549 +1,500 @@
+            }
+        }
+
+        if (normalSource_.get() == NormalSource::GenerateVertex) {
+            if (!copy) copy = std::shared_ptr<Mesh>(mesh->clone());
+            meshutil::calculateMeshNormals(*copy, normalComputationMode_);
+        }
+        enhancedMeshes_.push_back(copy ? copy : mesh);
+    }
+}
+
+MeshRasterization::MeshRasterization(const MeshRasterizer& processor)
+    : enhancedMeshes_(processor.enhancedMeshes_)
+    , forceOpaque_(processor.forceOpaque_)
+    , showFace_{processor.faceSettings_[0].show_, processor.faceSettings_[1].show_}
+    , tfTextures_{processor.faceSettings_[0].transferFunction_->getData(),
+                  processor.faceSettings_[processor.faceSettings_[1].sameAsFrontFace_.get() ? 0 : 1]
+                      .transferFunction_->getData()}
+    , shader_(processor.shader_) {}
+
+void MeshRasterizer::initializeResources() {
+    auto fso = shader_->getFragmentShaderObject();
+
+    fso->addShaderExtension("GL_NV_gpu_shader5", true);
+    fso->addShaderExtension("GL_EXT_shader_image_load_store", true);
+    fso->addShaderExtension("GL_NV_shader_buffer_load", true);
+    fso->addShaderExtension("GL_EXT_bindable_uniform", true);
+
+    // shading defines
+    utilgl::addShaderDefines(*shader_, lightingProperty_);
+
+    const std::array<std::pair<std::string, bool>, 15> defines = {
+        {{"USE_FRAGMENT_LIST", !forceOpaque_.get()},
+         {"ALPHA_UNIFORM", alphaSettings_.enableUniform_},
+         {"ALPHA_ANGLE_BASED", alphaSettings_.enableAngleBased_},
+         {"ALPHA_NORMAL_VARIATION", alphaSettings_.enableNormalVariation_},
+         {"ALPHA_DENSITY", alphaSettings_.enableDensity_},
+         {"ALPHA_SHAPE", alphaSettings_.enableShape_},
+         {"DRAW_EDGES", faceSettings_[0].showEdges_ || faceSettings_[1].showEdges_},
+         {"DRAW_EDGES_DEPTH_DEPENDENT", edgeSettings_.depthDependent_},
+         {"DRAW_EDGES_SMOOTHING", edgeSettings_.smoothEdges_},
+         {"MESH_HAS_ADJACENCY", meshHasAdjacency_},
+         {"DRAW_SILHOUETTE", drawSilhouette_},
+         {"SEND_TEX_COORD", faceSettings_[0].hatching_.hatching_.isChecked() ||
+                                faceSettings_[1].hatching_.hatching_.isChecked()},
+         {"SEND_SCALAR", faceSettings_[0].colorSource_ == ColorSource::TransferFunction ||
+                             faceSettings_[1].colorSource_ == ColorSource::TransferFunction},
+         {"SEND_COLOR", faceSettings_[0].colorSource_ == ColorSource::VertexColor ||
+                            faceSettings_[1].colorSource_ == ColorSource::VertexColor}}};
+
+    for (auto&& [key, val] : defines) {
+        for (auto&& so : shader_->getShaderObjects()) {
+            so.setShaderDefine(key, val);
+        }
+    }
+
+    shader_->build();
+}
+
+void MeshRasterization::rasterize(const ivec2& imageSize,
+                                  std::function<void(Shader&)> setUniformsRenderer) const {
+
+    shader_->activate();
+    if (!shader_->isReady()) return;

```

Comment:
This should be checked before it is activated?

Reply:
Rather not checked, will throw error, which is what we want.